src/Intl/Idn/Punycode/Punycode.php

```diff
@@ -0,0 +1,371 @@
+        $n = self::INITIAL_N;
+        $delta = 0;
+        $out = 0;
+        $maxOut = null === $outLength ? self::MAX_INT : $outLength;
+        $bias = self::INITIAL_BIAS;
+        $inputLength = 0;
+        $output = '';
+        $iter = new CodePointString($input);
+
+        foreach ($iter as $j => $codePoint) {
+            ++$inputLength;
+
+            if ($codePoint < 0x80) {
+                if ($maxOut - $out < 2) {
+                    throw new OutputSizeExceededException();
+                }
+
+                $output .= isset($caseFlags[$j])
+                    ? self::encodeBasic($codePoint, $caseFlags[$j])
+                    : \chr($codePoint);
+                ++$out;
+            }
+        }
+
+        $h = $out;
+        $b = $out;
+
+        if ($b > 0) {
+            $output .= self::DELIMITER;
+            ++$out;
+        }
+
+        while ($h < $inputLength) {
+            $m = self::MAX_INT;
+
+            foreach ($iter as $codePoint) {
+                if ($codePoint >= $n && $codePoint < $m) {
+                    $m = $codePoint;
+                }
+            }
+
+            if ($m - $n > \intdiv(self::MAX_INT - $delta, $h + 1)) {
+                throw new OverflowException();
+            }
+
+            $delta += ($m - $n) * ($h + 1);
+            $n = $m;
+
+            foreach ($iter as $j => $codePoint) {
+                if ($codePoint < $n && 0 === ++$delta) {
+                    throw new OverflowException();
+                } elseif ($codePoint === $n) {
+                    $q = $delta;
+
+                    for ($k = self::BASE; /* no condition */; $k += self::BASE) {
+                        if ($out >= $maxOut) {
+                            throw new OutputSizeExceededException();
+                        }
+
+                        if ($k <= $bias) {
+                            $t = self::TMIN;
+                        } elseif ($k >= $bias + self::TMAX) {
+                            $t = self::TMAX;
+                        } else {
+                            $t = $k - $bias;
+                        }
+
+                        if ($q < $t) {
+                            break;
+                        }
+
+                        $qMinusT = $q - $t;
+                        $baseMinusT = self::BASE - $t;
+                        $output .= self::encodeDigit($t + ($qMinusT) % ($baseMinusT), false);
+                        ++$out;
+                        $q = \intdiv($qMinusT, $baseMinusT);
+                    }
+
+                    $output .= self::encodeDigit($q, isset($caseFlags[$j]) ? $caseFlags[$j] : false);
+                    ++$out;
+                    $bias = self::adaptBias($delta, $h + 1, $h === $b);
+                    $delta = 0;
+                    ++$h;
+                }
+            }
+
+            ++$delta;
+            ++$n;
+        }
+
+        return $output;
+    }
+
+    /**
+     * @codeCoverageIgnore
+     */
+    private function __construct()
```

Comment:
I'd suggest removing all those empty constructors and adding `@internal` annotations on all classes

Reply:
Done